### ISSUE TITLE
fork/exec ./app/node_modules/electron/dist/electron: no such file or directory

### ISSUE THREAD

**Issue by twlz0ne** [Equanox/gotron#9]
Behaviour:

```sh
⋊> go run .
Loaded Configuration:
server:127.0.0.1
port:9109
name:Gotron
---
!prod
2018/10/29 15:12:02 Starting Electron...
2018/10/29 15:12:02 fork/exec ./app/node_modules/electron/dist/electron: no such file or directory
exit status 1
```

Only `Electron.app` in the `dist/` directory:

```sh
⋊> ls app/node_modules/electron/dist/
Electron.app           LICENSES.chromium.html
LICENSE                version
```

- macOS 10.11.6
- gotron 7c479e1 (2018-10-09)

**Comment by Equanox** [comment]
Thx, for opening this issue.
Does #10 fix it?

**Comment by twlz0ne** [comment]
```sh
⋊> go run .
Loaded Configuration:
port:9109
name:Gotron
server:127.0.0.1
---
!prod
2018/10/30 01:16:14 Starting Electron...
2018/10/30 01:16:14 fork/exec ./app/node_modules/electron/dist/Electron.app: permission denied
exit status 1
```

Actually, `Electron.app` is  a folder, we need `open` command to launch it:

```sh
⋊> open app/node_modules/electron/dist/Electron.app --args file://$PWD/app/assets/index.html
```

I aslo did a test in go:

```sh
⋊> cat test-exec.go
package main

import (
	"log"
	"os"
	"os/exec"
)

func main() {
	log.Printf("Starting Electron...")
	pwd, err := os.Getwd()
	if err != nil {
		log.Fatal(err)
	}
	path, exe, args := "./app/node_modules/electron/dist/", "Electron.app", "file://"+pwd+"/app/assets/index.html"
	cmd := exec.Command("open", path+exe, "--args", args)
	err = cmd.Start()
	if err != nil {
		log.Fatal(err)
	}
	log.Printf("%s succesfully started", exe)
}

⋊> go run test-exec.go
2018/10/30 03:11:36 Starting Electron...
2018/10/30 03:11:36 Electron.app succesfully started
```

![screenshot_2018-10-30_at_3 15 00_am-shadow_ 3 3-0-0 15 00_am](https://user-images.githubusercontent.com/2653486/47674483-5df97c00-dbf2-11e8-9743-327a6bf65433.png)

**Comment by Equanox** [comment]
Thx for the help. Updated #10 as proposed. Just referenced the app folder instead of the index.html .
Can you check if this fixes your problem?

**Comment by twlz0ne** [comment]
@Equanox Must be an absolute path:

![screenshot_2018-11-02_at_12 18 15_am](https://user-images.githubusercontent.com/2653486/47864588-f3844e00-de34-11e8-9e9b-b1472b94b07e.png)

**Comment by Equanox** [comment]
All right, :crossed_fingers:  . #10

**Comment by twlz0ne** [comment]
<del>

```diff
@@ -17,7 +17,7 @@ func FrontendPath() (dir string, exe string, args string, err error) {
        if err != nil {
            return "", "", "", err
        }
-       return "./app/node_modules/electron/dist/", "Electron.app", appDirectory, nil
+       return "./app/node_modules/electron/dist/", "Electron.app", "file://"+appDirectory+"/assets/index.html", nil
    }

    return "./app/node_modules/electron/dist/", "electron", "app", nil
```

</del>

**Comment by Equanox** [comment]
From `electron --help`
```
  A path to an Electron app may be specified. It must be one of the following:
    - index.js file.
    - Folder containing a package.json file.
    - Folder containing an index.js file.
    - .html/.htm file.
    - http://, https://, or file:// URL.
```
In `./app/package.json` we have defined `main.js` as our startup point. From there we include `./app/assets/index.html` and rampup the websocket communication between go<=>nodejs. So we can not just use the `./app/assets/index.html`, we have to rampup electron using a `Folder containing a package.json file.` 

So #10  doesn't work as intended?

**Comment by twlz0ne** [comment]
Oh, my mistake, I ignored an error before:

```sh
⋊> po npm run install:app

> gotron@0.0.1 install:app ~/repos/go/gui/gotron
> cd app && npm install && npm run rebuild

npm WARN electron-rampup@0.1.0 No description
npm WARN electron-rampup@0.1.0 No repository field.
npm WARN electron-rampup@0.1.0 No license field.

audited 512 packages in 25.576s
found 0 vulnerabilities


> electron-rampup@0.1.0 rebuild ~/repos/go/gui/gotron/app
> run-script-os


> electron-rampup@0.1.0 rebuild:darwin ~/repos/go/gui/gotron/app
> electron-rebuild.app

sh: electron-rebuild.app: command not found
npm ERR! file sh
npm ERR! code ELIFECYCLE
npm ERR! errno ENOENT
npm ERR! syscall spawn
npm ERR! electron-rampup@0.1.0 rebuild:darwin: `electron-rebuild.app`
npm ERR! spawn ENOENT
npm ERR!
npm ERR! Failed at the electron-rampup@0.1.0 rebuild:darwin script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     ~/.npm/_logs/2018-11-08T03_55_53_299Z-debug.log
```

There's no `electron-rebuild.app` on Mac, after making the following modifications, the error disappears:

```diff
⋊> git log -n1 --oneline
4255ab7 Using absolute path of app folder on macOS

⋊> git diff -U0
────────────────────────────────────────────────────────────────────────────────
modified: app/package.json
────────────────────────────────────────────────────────────────────────────────
@@ -16 +16 @@
-    "rebuild:darwin": "./node_modules/.bin/electron-rebuild.app",
+    "rebuild:darwin": "./node_modules/.bin/electron-rebuild",
```

No need pass `file://.../index.html` to `--args`.

**Comment by puengel** [comment]
This issue won't be fixed in gotron boilerplate.
The issue has been fixed and tested in the new gotron rework #14 
[Release v0.2.0](https://github.com/Equanox/gotron/tree/v0.2.0)
See latest [Readme](https://github.com/Equanox/gotron/blob/master/README.md) for more information on how to use the new gotron api.
